### PR TITLE
Split ext_clear_storage out from ext_set_storage

### PR DESCRIPTION
**Pull request by Robbepop**
This PR splits the
```
ext_set_storage(key_ptr: u32, value_non_null: u32, value_ptr: u32, value_len: u32);
```
into two APIs:
```
ext_set_storage(key_ptr: u32, value_ptr: u32, value_len: u32);
ext_clear_storage(key_ptr: u32);
```

## Motivation

We perform this change to make it simpler for static analyzers to analyze the underlying code semantically.

The old API with its `value_non_null` parameter is theoretically impossible to analyze statically since the `value_non_null` parameter could be any runtime computed value. So it is potentially very hard for static analyzers to apply the correct gas costs in the case where we want to differentiate between setting and clearing contract storage entries.

**Why do we want to differentiate between those two operations?**

The operations have completely different semantics:
- The one API is updating a value of a storage entry,
- the other is clearing the value of a storage entry.

Also as described above this makes it far simpler for static analyzers (or in some cases even possible) to reason about the underlying code. It slightly reduces overhead in some situations but that should generally be negligible.

## Open Questions

While working on this PR one particular question arosed:

How do we currently treat empty storage entries? E.g. when setting `ext_set_storage` with a `value_len` of 0? This semantically is different from having a cleaned-up contract storage entry and thus should be taken into account for rental fees.

Since we were not aware of any smart contract use cases that could even profit from creating or having occupied empty storage entries we also disallowed having them in the adjusted `ext_set_storage`. Upon trying to set a contract storage entry to be empty this function will trigger a host error and the contract will trap.

We might change the semantics of this particular scenario once we are aware of a proper smart contracts use case.

## Task List of this PR

- [x] Introduced `ext_clear_storage`
- [x] Adjusted semantics of `ext_set_storage`
- [x] Adjusted `COMPLEXITY.md`
- [x] Adjusted SEAL tests

### CODE REVIEW

File: frame/contracts/src/wasm/runtime.rs [hunk 1]
```diff
@@ -348,30 +348,49 @@ define_env!(Env, <E: Ext>,
...
+	//
+	// - If value length exceeds the configured maximum value length of a storage entry.
+	// - Upon trying to set an empty storage entry (value length is 0).
+	ext_set_storage(ctx, key_ptr: u32, value_ptr: u32, value_len: u32) => {
+		if value_len > ctx.ext.max_value_size() {
+			// Bail out if value length exceeds the set maximum value size.
+			return Err(sp_sandbox::HostError);
+		}
+		if value_len == 0 {
+			// Bail out on setting storage to an emptry entry.
```
Comment (pepyakin) [inline frame/contracts/src/wasm/runtime.rs]:
```suggestion
			// Bail out on setting storage to an empty entry.
```

File: frame/contracts/src/wasm/runtime.rs [hunk 2]
```diff
@@ -348,30 +348,49 @@ define_env!(Env, <E: Ext>,
...
+	// # Errors
+	//
+	// - If value length exceeds the configured maximum value length of a storage entry.
+	// - Upon trying to set an empty storage entry (value length is 0).
+	ext_set_storage(ctx, key_ptr: u32, value_ptr: u32, value_len: u32) => {
+		if value_len > ctx.ext.max_value_size() {
+			// Bail out if value length exceeds the set maximum value size.
+			return Err(sp_sandbox::HostError);
+		}
+		if value_len == 0 {
```
Comment (pepyakin) [inline frame/contracts/src/wasm/runtime.rs]: Let's defer this to #5107

**Comment by gavofyork** [comment]
The proposition that this PR is based on seems flawed. Storage rent should be all the incentivisation needed to ensure that storage clean-ups occur. Trying to warp the prices away from the true costs for the storage operations in order to try to incentivise cleaning up or disincentivise wasteful usage seems accordingly ill-considered.

**Comment by Robbepop** [comment]
> The proposition that this PR is based on seems flawed. Storage rent should be all the incentivisation needed to ensure that storage clean-ups occur. Trying to warp the prices away from the true costs for the storage operations in order to try to incentivise cleaning up or disincentivise wasteful usage seems accordingly ill-considered.

You are right that the original proposition was flawed since we would never decouple gas prices of operations from their real performance impact on the system which the original proposition was based on. - Unfortunately I forgot to update after we had internal discussions about this. However, the code changes no longer reflect the original proposition.
However, there are other reasons for splitting this API up into 2 functions because they are semantically doing completely different things (free and set) and by having this split we can more easily analyze semantics of the underlying code statically.

**Pull request merged**